odl/ufunc_ops/ufunc_ops.py

```diff
@@ -0,0 +1,196 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Ufunc operators for ODL vectors."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+from odl.set.space import LinearSpace
+from odl.space import ProductSpace, fn
+from odl.operator import Operator, MultiplyOperator
+from odl.util.utility import is_int_dtype
+from odl.util.ufuncs import UFUNCS
+
+__all__ = ()
+
+
+def _is_integer_only_ufunc(name):
+    return 'shift' in name or 'bitwise' in name or name == 'invert'
+
+LINEAR_UFUNCS = ['conj', 'negative', 'rad2deg', 'deg2rad', 'add', 'subtract']
+
+RAW_INIT_DOCSTRING = """
+
+"""
+
+
+RAW_EXAMPLES_DOCSTRING = """
+Examples
+--------
+>>> import odl
+>>> space = odl.{space}
+>>> op = {name}(space)
+>>> print(op({arg}))
+{result}
+"""
+
+
+class UfuncOperator(Operator):
+
+    """Base class for all ufunc operators."""
+
+
+def derivative_factory(name):
+    """Create derivative function for some ufuncs."""
+
+    if name == 'sin':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            point = self.domain.element(point)
+            return MultiplyOperator(point.ufunc.cos())
+        return derivative
+    elif name == 'cos':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            point = self.domain.element(point)
+            return MultiplyOperator(-point.ufunc.sin())
+        return derivative
+    elif name == 'tan':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            return MultiplyOperator(1 + self(point) ** 2)
+        return derivative
+    elif name == 'sqrt':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            return MultiplyOperator(0.5 / self(point))
+        return derivative
+    elif name == 'square':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            point = self.domain.element(point)
+            return MultiplyOperator(2.0 * point)
+        return derivative
+    elif name == 'log':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            point = self.domain.element(point)
+            return MultiplyOperator(1.0 / point)
+        return derivative
+    elif name == 'exp':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            return MultiplyOperator(self(point))
+        return derivative
+    else:
+        # Fallback to default
+        return UfuncOperator.derivative
+
+
+def ufunc_class_factory(name, nargin, nargout, docstring):
+    """Create a UfuncOperator from a given specification."""
+
+    assert 0 <= nargin <= 2
```

Comment:
this is the range of values currently supported and currently available in numpy. I simply added it here in case numpy adds more ufuncs in the future so we don't forget to update it.